### ISSUE TITLE
Improve/replace `omxDefaultComputePlan()`?

### ISSUE THREAD

**Issue by RMKirkpatrick** [OpenMx/OpenMx#145]
I wrote [`omxDefaultComputePlan()`](https://openmx.ssri.psu.edu/docs/OpenMx/latest/_static/Rdoc/omxDefaultComputePlan.html) to make it easier for users to create custom compute plans.  The idea was to have a function that would return what the default compute plan would be, given the function's argument values and the global values of mxOptions; the user could then modify the plan for his/her purposes.  Such a function would make it easy to give users simple instructions on how to make and use a particular custom compute plan.

But unfortunately, the function has some things missing.  For instance, there's no way to tell it that it's going to be used in an MxModel with MxConstraints.  For that, adding a `hasConstraints` argument would be an easy fix.  The only reason I haven't done so is because we'll have SEs in the presence of constraints in the foreseeable future, but we still haven't determined what the default compute plan should be for MxConstrained models after those SEs are implemented (parenthetically, I anticipate that the only step in the default plan that would be dropped in the presence of MxConstraints would be MxComputeHessianQuality).

Ideally, the user would be able to use `omxDefaultComputePlan()` to get a compute plan that could go into an MxModel with thresholds, and the user would be able to run that model and get the same behavior from `mxRun()` as in the case of no user-provided compute plan.  Unfortunately, adding a `hasThresholds` argument won't be enough.  Look at the code starting at what is currently line 160 of R/MxRun.R .  If the MxModel has any expectations, then the gradient step size and gradient iterations for the MxComputeNumericDeriv step, as well as the function precision (which is NPSOL-specific), are given values based on whether the MxModel contains thresholds (for additional relevant code, see the definition of `getPrecisionPerExpectation()` in R/MxExpectation.R, and the definition of `imxAutoOptionValue()` in R/MxOptions.R ).  The adjustment of the function precision is something that cannot be conveyed in a compute plan.

One possible solution is to make the function precision something that can be passed to NPSOL through the MxComputeGradientDescent step, via an argument to the step's constructor function.  Another solution would be to deprecate `omxDefaultComputePlan()`, and replace it with a similar function that requires an MxModel object to be populated with a default compute plan and an appropriate function precision.  The latter solution also has the advantage of automatically making the function aware of any mxOptions set locally to the MxModel, plus any thresholds or constraints in the model.  Or maybe custom compute plans are an advanced enough feature that none of this is worth worrying about.

Thoughts?

**Comment by jpritikin** [comment]
On Wed, Feb 27, 2019 at 11:11:43AM -0800, RMKirkpatrick wrote:
>    Another solution would be to deprecate omxDefaultComputePlan(), and 
>    replace it with a similar function that requires an MxModel object 
>    to be populated with a default compute plan and an appropriate 
>    function precision.

Given that mxRun relies on omxDefaultComputePlan, any improvements are 
constrained by backward compatibility.

How about adding a new function, omxGuessComputePlan, that takes the 
MxModel. Thresholds and mxConstraints could be detected and handled 
appropriately. In addition, we could detect mxExpectationBA81 and 
produce a compute plan suitable for IFA models. Similarly, a compute 
plan could be automatically built to address the recently added 
recompute='onDemand' mxAlgebra option.

**Comment by mcneale** [comment]
So I figure the compute plan that was used by a model when it has been mxRun() is returned and can be extracted.  It is only un-run models that lack the plan.  While less generally useful, after-the-fact compute plans seem to avoid all the issues Rob raised.  We could even provide some example  here's-one-I-prepared-earlier fitted models for this purpose.

The other thought I had was whether processing could occur with no optimization or anything being calculated, but a walk-through sanity check of the model that could return the compute plan (even if each step had a "check-only" don't run flag.

**Comment by jpritikin** [comment]
Not sure, but I think the whole point of this issue is to produce a compute plan for a model that hasn't been run yet.

**Comment by mcneale** [comment]
I figured that,  but unless the default plan runs very slowly, it doesn't seem that big a deal to get the plan from the output model.  Unless it cannot run without a custom compute plan...

**Comment by RMKirkpatrick** [comment]
> Given that mxRun relies on omxDefaultComputePlan, any improvements are constrained by backward compatibility.

True, but changes to `omxDefaultComputePlan()` can still be made without breaking anything.

> How about adding a new function, omxGuessComputePlan, that takes the MxModel. Thresholds and mxConstraints could be detected and handled appropriately. In addition, we could detect mxExpectationBA81 and produce a compute plan suitable for IFA models. Similarly, a compute plan could be automatically built to address the recently added recompute='onDemand' mxAlgebra option.

I like that idea.
